### PR TITLE
Get rid of static building counts

### PR DESCRIPTION
**Pull request by crudelios**
Count the total/active buildings as needed instead.

This is possible because we can now cycle buildings by type, making the loops faster.

Remove the 10 dock limit since it was closely linked with static building counts.

Also fix some bugs, like a rioter only destroying a part of an hippodrome.

Create a new "visited buildings" list instead of the old bitfield.

This may add some bugs over the resource bugs.

### CODE REVIEW

**Review by PrettyFlower — CHANGES_REQUESTED** [review]
Ok so some of this was definitely beyond me but as mentioned down below, I'd like to try changing around the visited_buildings data structures before we merge this guy in.

File: src/figure/visited_buildings.c [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+#include "building/dock.h"
+#include "city/data_private.h"
+#include "core/array.h"
+#include "core/log.h"
+#include "figure/figure.h"
+
+#define VISITED_BUILDINGS_ARRAY_SIZE_STEP 100
+#define VISITED_BUILDINGS_BUFFER_SIZE (sizeof(int32_t) * 2)
+
+static array(visited_building) visited_buildings;
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/figure/visited_buildings.c:12]: Yes that's exactly what we want, but this way was the easiest one and it works. It's a bit bug prone, but my reasoning is "Nah, it'll be fine!"

File: src/city/data.c [hunk 2]
```diff
@@ -111,10 +113,9 @@ static void save_main_data(buffer *main)
     buffer_write_u8(main, city_data.map.exit_point.x);
     buffer_write_u8(main, city_data.map.exit_point.y);
     buffer_write_i16(main, city_data.map.exit_point.grid_offset);
-    buffer_write_u8(main, city_data.building.senate_x);
-    buffer_write_u8(main, city_data.building.senate_y);
-    buffer_write_i16(main, city_data.building.senate_grid_offset);
-    buffer_write_i32(main, city_data.building.senate_building_id);
+    for (int i = 0; i < 8; i++) {
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/city/data.c:116]: Well, if I let them go, then on the read state function we'll have a bunch more or `if (version)`s. It already looks bad add it is now...

File: src/building/dock.c [hunk 3]
```diff
@@ -296,21 +297,28 @@ static int destination_dock_ready_for_ship(figure *ship)
...
         return 0;
     }
 
-    handled_good handled_goods[MAX_DOCKS];
-    get_already_handled_goods(handled_goods, ship_id);
+    handled_good *handled_goods = malloc(sizeof(handled_good) * total_docks);
+    if (!handled_goods) {
+        return 0;
+    }
+    get_already_handled_goods(handled_goods, total_docks, ship_id);
```
[the remark this answers is not in the document]
Comment (crudelios) [reply src/building/dock.c:309]: I think here the relative simplicity of the code doesn't justify using a dynamic  array. However, we can set both `handled_goods` and `total_docks` as static variables with file scope,  preventing the need to pass either of them as function arguments.

File: src/figure/visited_buildings.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (crudelios) [reply src/figure/visited_buildings.c:12]: What about using this:

https://github.com/P-p-H-d/mlib

File: src/building/dock.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (PrettyFlower) [reply src/building/dock.c:309]: Oh I'd much rather pass them around than introduce more globals.

**Pull request merged**